pex/scie/model.py

```diff
@@ -33,86 +34,166 @@ class Value(Enum.Value):
+                add = self.add_all or dist.metadata.project_name in self.add_project
                 if not remove and not add:
                     continue
-                for script in dist.get_entry_map().get("console_scripts", {}):
+                for name, named_entry_point in (
+                    dist.get_entry_map().get("console_scripts", {}).items()
+                ):
                     if remove:
-                        console_scripts.discard(script)
+                        console_scripts.pop(ConsoleScript(name=name), None)
+                        console_scripts.pop(
+                            ConsoleScript(name=name, project_name=dist.metadata.project_name), None
+                        )
                     else:
-                        console_scripts.add(script)
-        for script in self.remove_individual:
-            console_scripts.discard(script)
-        return console_scripts
-
-
-@attr.s(frozen=True)
-class ModuleEntryPoint(object):
-    @classmethod
-    def try_parse(cls, value):
-        # type: (str) -> Optional[ModuleEntryPoint]
-        name, _, entry_point = value.partition(":")
-        return cls(name, entry_point) if entry_point else None
+                        console_scripts[
+                            ConsoleScript(name=name, project_name=dist.metadata.project_name)
+                        ] = named_entry_point
+
+        for console_script in self.remove_individual:
+            console_scripts.pop(console_script, None)
+            if not console_script.project_name:
+                for cs in tuple(console_scripts):
+                    if console_script.name == cs.name:
+                        console_scripts.pop(cs)
+
+        def iter_entry_points():
+            # type: () -> Iterator[NamedEntryPoint]
+            not_found = []  # type: List[ConsoleScript]
+            wrong_project = []  # type: List[Tuple[ConsoleScript, Distribution]]
+            for script, ep in console_scripts.items():
+                if ep:
+                    yield ep
+                else:
+                    dist_entry_point = get_entry_point_from_console_script(script.name, dists)
+                    if not dist_entry_point:
+                        not_found.append(script)
+                    elif (
+                        script.project_name
+                        and dist_entry_point.dist.metadata.project_name != script.project_name
+                    ):
+                        wrong_project.append((script, dist_entry_point.dist))
+                    else:
+                        yield NamedEntryPoint(
+                            name=dist_entry_point.name, entry_point=dist_entry_point.entry_point
+                        )
 
-    name = attr.ib()  # type: str
-    entry_point = attr.ib()  # type: str
+            if not_found or wrong_project:
+                raise ValueError(
+                    # TODO(John Sirois): XXX: Craft an error message.
```

Comment:
Reminder about this

Reply:
Fixed and added test coverage for missing / mis-placed scripts.